pallets/pooled-staking/src/calls.rs

```diff
@@ -0,0 +1,459 @@
+        // As with all pools there will be some rounding error, this amount
+        // should be small enough so that it is safe to directly release it
+        // in the delegator account.
+        let leaving_shares =
+            pools::Leaving::<T>::stake_to_shares_or_init(&candidate, removed_stake.clone())?;
+        let leaving_stake =
+            pools::Leaving::<T>::add_shares(&candidate, &delegator, leaving_shares.clone())?;
+        pools::Leaving::<T>::increase_hold(&candidate, &delegator, &leaving_stake)?;
+
+        // We create/mutate a request for leaving.
+        let now = frame_system::Pallet::<T>::block_number();
+        let operation_key = PendingOperationKey::Leaving {
+            candidate: candidate.clone(),
+            at_block: now,
+        };
+        let operation = PendingOperations::<T>::get(&delegator, &operation_key);
+        let operation = operation
+            .err_add(&leaving_shares.0)
+            .map_err(|_| Error::<T>::MathOverflow)?;
+        PendingOperations::<T>::set(&delegator, &operation_key, operation);
+
+        // We release the dust if non-zero.
+        let dust = removed_stake
+            .0
+            .err_sub(&leaving_stake.0)
+            .map_err(Error::<T>::from)?;
+
+        if dust.is_zero() {
+            T::Currency::release(
+                &T::CurrencyHoldReason::get(),
+                &delegator,
+                dust,
+                Precision::Exact,
+            )?;
+            Candidates::<T>::sub_total_stake(&candidate, Stake(dust))?;
+        }
+
+        Pallet::<T>::deposit_event(Event::<T>::RequestedUndelegate {
+            candidate,
+            delegator,
+            from: pool,
+            pending: leaving_stake.0,
+            released: dust,
+        });
+
+        Ok(().into())
+    }
+
+    pub fn execute_pending_operations(
+        origin: OriginFor<T>,
+        operations: Vec<PendingOperationQuery<T::AccountId, T::BlockNumber>>,
+    ) -> DispatchResultWithPostInfo {
+        // We don't care about the sender.
+        let _ = ensure_signed(origin)?;
+
+        for (index, query) in operations.into_iter().enumerate() {
+            // We deconstruct the query and find the balance associated with it.
+            // If it is zero it may not exist or have been executed before, thus
+            // we simply skip it instead of erroring.
+            let PendingOperationQuery {
+                delegator,
+                operation,
+            } = query;
+
+            let value = PendingOperations::<T>::get(&delegator, &operation);
+
+            if Zero::is_zero(&value) {
+                continue;
+            }
+
+            match &operation {
+                PendingOperationKey::JoiningAutoCompounding {
+                    candidate,
+                    at_block,
+                } => {
+                    ensure!(
+                        T::JoiningRequestFilter::can_be_executed(&candidate, &delegator, *at_block),
```

Comment:
Is the goal that we set JoiningRequestFilter to have the delay of at least 1 session?

Reply:
Exactly. This trait allows to use delays in blocks or session boundaries. We may remove `candidate` and `delegator` arguments tho.